File: medium/238.ts
/**
    238. Product of Array Except Self
 */
function productExceptSelf(nums: number[]): number[] {
    const n: number = nums.length
    const pf: number[] = new Array(n + 1).fill(0)
    const sf: number[] = new Array(n + 1).fill(0)
    pf[0] = 1
    sf[n] = 1
    for (let i = 1; i <= n; ++i)
        pf[i] = pf[i - 1] * nums[i - 1]
    for (let i = n - 1; i >= 0; --i)
        sf[i] = sf[i + 1] * nums[i]
    const res: number[] = []
    for (let i = 1; i <= n; ++i)
        res.push(pf[i - 1] * sf[i])

    return res
};

File: medium/238.java
/**
    # 238. Product of Array Except Self
 */
class Solution {
    public int[] productExceptSelf(int[] nums) {
        final int n = nums.length;
        int[] prf = new int[n];
        int[] suf = new int[n];
        int[] res = new int[n];

        prf[0] = nums[0];
        suf[n - 1] = nums[n - 1];

        for(int i = 1; i < n; ++i)
            prf[i] = prf[i - 1] * nums[i];
        for(int i = n - 2; i >= 0; --i)
            suf[i] = suf[i + 1] * nums[i];
        for(int i = 1; i < n - 1; ++i)
            res[i] = prf[i - 1] * suf[i + 1];
        
        res[0] = suf[1];
        res[n - 1] = prf[n - 2];

        return res;
    }
}
